### DISCUSSION TITLE
Allow excluding directories from Borg backup

### DISCUSSION THREAD

**Discussion by Gchbg** [nextcloud/all-in-one#4751] [Ideas] [answered]
### Is your feature request related to a problem? Please describe. 

My AIO instance has limited storage. When my files are backed up locally by the borg system, my effective storage is halved. Some of my data can be regenerated if lost, or is otherwise not critical, and I would like to exlude it from backup.

### Describe the solution you'd like

I would love to have an easy way to exclude certain directories from all backups. This can be easily achieved by adding two command-line options to the `borg create` invocation in the aio-borgbackup container: `--exclude-if-present=.noaiobackup --keep-exclude-tags`. This will exclude any directory which contains a file called `.noaiobackup` within it.

In this way, I can exclude directories with non-critical data from my backups, reducing the storage requirement for my local backup, and making more of my limited storage available for actual data.

I've been using this solution on a locally modified AIO instance without issues.

### Additional context

This change is trivial and completely backward-compatible. People who do not have `.noaiobackup` files will not experience any difference.

**Comment by szaimen** [comment]
Hi, interesting idea! Just to make it clear directly, I will not work on this but would accept a PR that adds this feature. 

Please make sure that the restore with rsync also respects these files then so that it will not delete all the folders that contain these files when you try to do a restore from the AIO interface.

**Comment by szaimen** [comment, marked as answer]
This will be done in https://github.com/nextcloud/all-in-one/pull/6009

**Reply by szaimen** [reply]
This is now released with v10.6.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by Iakhou69** [comment]
Hi, so now we just need to add this file : .noaiobackup and that's all ? Or we need to add this on borg container ? --exclude-if-present=.noaiobackup --keep-exclude-tags

**Reply by szaimen** [reply]
See https://github.com/nextcloud/all-in-one?tab=readme-ov-file#how-to-exclude-nextclouds-data-directory-or-the-preview-folder-from-backup

**Comment by Asuranceturix** [comment]
Is it possible to exclude arbitrary folders from backup? We have a group folder that is used to temporarily exchange files between users, and it makes no sense to include it in the backup. I have created a `.noaiobackup` file inside that folder, but Borg does include it in the backup.

**Reply by szaimen** [reply]
Indeed this is currently not possible. See what is possible: https://github.com/nextcloud/all-in-one?tab=readme-ov-file#how-to-exclude-nextclouds-data-directory-or-the-preview-folder-from-backup

**Reply by Asuranceturix** [reply]
Well, too bad, it would be nice to have but I guess it's kind of a corner use case. Anyhow, thanks for your extremely prompt reply and for your work on the project. :)

**Comment by JMoVS** [comment]
Hi @szaimen,

I just dug into the code a bit of the pull request that solved this issue, and I am wondering why here the data directory is hard coded to the standard volume based one and not also the arbitrary one?

https://github.com/nextcloud/all-in-one/pull/6009/commits/13afcd944ada6c754ddd784975fe8b9d7c504b1a#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR200-R201

Another question: If I understand this correctly, what is currently done is that only the .noaiobackup file is included in the backup but the rest is not?

Last but not least: Furthermore, why is this check manual and limited to these two folders? Or what is the rational against using the borg option as initially proposed `--exclude-if-present=.noaiobackup --keep-exclude-tags`?

I am considering to volunteer a pull request for the changes to the code and the documentation to land that properly as I'm currently running into this issues as a limitation of my setup :)

**Reply by szaimen** [reply]
> Hi @szaimen,

Hey 👋

> I just dug into the code a bit of the pull request that solved this issue, and I am wondering why here the data directory is hard coded to the standard volume based one and not also the arbitrary one?
> 
> [13afcd9#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR200-R201](https://github.com/nextcloud/all-in-one/commit/13afcd944ada6c754ddd784975fe8b9d7c504b1a#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR200-R201)

This is because any configured target NEXTCLOUD_DATADIR always gets mounted in `/nextcloud_aio_volumes/nextcloud_aio_nextcloud_data` inside the backup container. See https://github.com/nextcloud/all-in-one/blob/a9d462489c5fa7f568e8ced174177d565662430e/php/containers.json#L548-L549

> Another question: If I understand this correctly, what is currently done is that only the .noaiobackup file is included in the backup but the rest is not?

Yes, correct.

> Last but not least: Furthermore, why is this check manual and limited to these two folders? Or what is the rational against using the borg option as initially proposed `--exclude-if-present=.noaiobackup --keep-exclude-tags`?

The problem is that we use rsync with the `--delete` flag for restoring from local backup which deletes empty directories. We would manually implement the same behaviour of borg for rsync but I have not found a way to do so. This is why the `.noaiobackup` feature is only implemented for two sensible locations for now.

**Reply by JMoVS** [reply]
> > Hi @szaimen,
> 
> Hey 👋
> 
> > I just dug into the code a bit of the pull request that solved this issue, and I am wondering why here the data directory is hard coded to the standard volume based one and not also the arbitrary one?
> > [13afcd9#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR200-R201](https://github.com/nextcloud/all-in-one/commit/13afcd944ada6c754ddd784975fe8b9d7c504b1a#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR200-R201)
> 
> This is because any configured target NEXTCLOUD_DATADIR always gets mounted in `/nextcloud_aio_volumes/nextcloud_aio_nextcloud_data` inside the backup container. See
> 
> https://github.com/nextcloud/all-in-one/blob/a9d462489c5fa7f568e8ced174177d565662430e/php/containers.json#L548-L549
> 
interesting, thanks!

> > Another question: If I understand this correctly, what is currently done is that only the .noaiobackup file is included in the backup but the rest is not?
> 
> Yes, correct.
> 
> > Last but not least: Furthermore, why is this check manual and limited to these two folders? Or what is the rational against using the borg option as initially proposed `--exclude-if-present=.noaiobackup --keep-exclude-tags`?
> 
> The problem is that we use rsync with the `--delete` flag for restoring from local backup which deletes empty directories. We would manually implement the same behaviour of borg for rsync but I have not found a way to do so. This is why the `.noaiobackup` feature is only implemented for two sensible locations for now.

Have you looked at using the rsync directpry merge filter option? `--filter=':- .noaiobackup'` or the rsync -F option, from the man page:
```
        -F                       same as --filter='dir-merge /.rsync-filter'
                                repeated: --filter='- .rsync-filter'
```

If I understand that option correctly, it wouldn't delete the directories as the directories then simply wouldn't be considered

**Reply by szaimen** [reply]
> Have you looked at using the rsync directpry merge filter option? `--filter=':- .noaiobackup'` or the rsync -F option, from the man page:
> 
> ```
>         -F                       same as --filter='dir-merge /.rsync-filter'
>                                 repeated: --filter='- .rsync-filter'
> ```
> 
> If I understand that option correctly, it wouldn't delete the directories as the directories then simply wouldn't be considered

I have not but I will also not dive into this again as the current solution is good enough imho.

However feel free to do it yourself :)

**Reply by JMoVS** [reply]
@szaimen Would you accept a PR if I volunteered it?

**Reply by szaimen** [reply]
Yes

**Reply by JMoVS** [reply]
Awesome I'll have a look at it. Do you have an easy way or automated way to test?

**Reply by szaimen** [reply]
> Awesome I'll have a look at it. Do you have an easy way or automated way to test?

Not really. The closest might be this: https://github.com/nextcloud/all-in-one/blob/main/.github/workflows/playwright.yml

**Reply by JMoVS** [reply]
Further clarification question:

During restore, would it be sufficient to *only* check in the backup archive for presence of .noaiobackup files (those would now be preserved in the backip phase) and therefore *don't* check the restore target system directory for the presence of a .noaiobackup file?

I only see one scenario where somebody has either unfortunate timing with the one time upgrade logic or someone is running a live system and restores a backip and mistakenly thinks if they put a .noaiobackup file into one of their directories and start a *restore*, that then their files would be left alone - but that seems something that could be handled with documentation.

Anyway, let me know what you think - for me the most logical would be to say that .noaiobackup controls only what is *backup up*, not what is restored. If the directory was not included in the backup, it will not be restored and no data will be overwritten

**Reply by szaimen** [reply]
> for me the most logical would be to say that .noaiobackup controls only what is _backup up_, not what is restored. If the directory was not included in the backup, it will not be restored and no data will be overwritten

I fear this will not work due to the usage of the delete flag: https://github.com/nextcloud/all-in-one/blob/7c57a3f702886504288f36bd48432774fd828c04/Containers/borgbackup/backupscript.sh#L393. However we need that to make sure that backup and restore target have the same state as otherwise the restore might brick the instance.

**Reply by JMoVS** [reply]
@szaimen Hi, I'm toying around here: https://github.com/JMoVS/all-in-one/tree/enh/noaiobackup-arbitrary-dirs - what is a good way of testing this? Spin up a VPS and backup and restore a bit to see if it works properly? I used this as a chance to toy around with claude code so I want to avoid you running tests to run into basic syntax errors that I should fix beforehand. I reviewed the code manually and think it does what I want it to do but happy to hear your perspective as well of course

**Reply by szaimen** [reply]
> what is a good way of testing this? Spin up a VPS and backup and restore a bit to see if it works properly?

Yes, would say so
